File: backend/src/main/java/br/ufpr/sept/so2/modules/solicitacoes/infrastructure/persistence/SolicitacaoJpaAdapter.java
package br.ufpr.sept.so2.modules.solicitacoes.infrastructure.persistence;

import br.ufpr.sept.so2.modules.solicitacoes.application.ports.SolicitacaoRepository;
import br.ufpr.sept.so2.modules.solicitacoes.domain.Solicitacao;
import br.ufpr.sept.so2.modules.solicitacoes.domain.SolicitacaoEvento;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Component
public class SolicitacaoJpaAdapter implements SolicitacaoRepository {

    private final SolicitacaoJpaRepository jpaRepository;
    private final SolicitacaoEventoJpaRepository eventoRepository;

    public SolicitacaoJpaAdapter(
            SolicitacaoJpaRepository jpaRepository,
            SolicitacaoEventoJpaRepository eventoRepository
    ) {
        this.jpaRepository = jpaRepository;
        this.eventoRepository = eventoRepository;
    }

    @Override
    public Solicitacao save(Solicitacao solicitacao) {
        SolicitacaoJpaEntity entity = jpaRepository.findById(solicitacao.getId())
                .orElseGet(() -> SolicitacaoJpaEntity.fromDomain(solicitacao));
        entity.merge(solicitacao);
        jpaRepository.save(entity);
        for (SolicitacaoEvento evento : solicitacao.getEventos()) {
            if (!eventoRepository.existsById(evento.id())) {
                eventoRepository.save(SolicitacaoEventoJpaEntity.fromDomain(solicitacao.getId(), evento));
            }
        }
        return carregar(solicitacao.getId()).orElseThrow();
    }

    @Override
    public Optional<Solicitacao> findById(UUID id) {
        return carregar(id);
    }

    @Override
    public Optional<Solicitacao> findByProtocolo(String protocolo) {
        return jpaRepository.findByProtocolo(protocolo)
                .map(entity -> entity.toDomain(eventosDe(entity.getId())));
    }

    @Override
    public Page<Solicitacao> findMinhas(
            UUID solicitanteId,
            String estado,
            String tipoCodigo,
            Integer ano,
            Pageable pageable
    ) {
        String anoPrefixo = ano == null ? null : "PROT-" + ano + "-%";
        return jpaRepository.findMinhas(solicitanteId, estado, tipoCodigo, anoPrefixo, pageable)
                .map(SolicitacaoJpaEntity::toDomainSemEventos);
    }

    private Optional<Solicitacao> carregar(UUID id) {
        return jpaRepository.findById(id).map(entity -> entity.toDomain(eventosDe(id)));
    }

    private List<SolicitacaoEvento> eventosDe(UUID solicitacaoId) {
        return eventoRepository.findBySolicitacaoIdOrderByCreatedAtDesc(solicitacaoId).stream()
                .map(SolicitacaoEventoJpaEntity::toDomain)
                .toList();
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/solicitacoes/infrastructure/persistence/SolicitacaoJpaAdapter.kt
package br.ufpr.sept.so2.modules.solicitacoes.infrastructure.persistence

import br.ufpr.sept.so2.modules.solicitacoes.application.ports.SolicitacaoRepository
import br.ufpr.sept.so2.modules.solicitacoes.domain.Solicitacao
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import java.util.Optional
import java.util.UUID

@Component
class SolicitacaoJpaAdapter(
    private val jpaRepository: SolicitacaoJpaRepository,
    private val eventoRepository: SolicitacaoEventoJpaRepository,
) : SolicitacaoRepository {

    override fun save(solicitacao: Solicitacao): Solicitacao {
        val entity = jpaRepository.findById(solicitacao.id)
            .orElseGet { SolicitacaoJpaEntity.fromDomain(solicitacao) }
        entity.merge(solicitacao)
        jpaRepository.save(entity)
        for (evento in solicitacao.eventos) {
            if (!eventoRepository.existsById(evento.id)) {
                eventoRepository.save(SolicitacaoEventoJpaEntity.fromDomain(solicitacao.id, evento))
            }
        }
        return carregar(solicitacao.id).orElseThrow()
    }

    override fun findById(id: UUID): Optional<Solicitacao> = carregar(id)

    override fun findByProtocolo(protocolo: String): Optional<Solicitacao> =
        jpaRepository.findByProtocolo(protocolo)
            .map { entity -> entity.toDomain(eventosDe(entity.id!!)) }

    override fun findMinhas(
        solicitanteId: UUID,
        estado: String?,
        tipoCodigo: String?,
        ano: Int?,
        pageable: Pageable,
    ): Page<Solicitacao> {
        val anoPrefixo = if (ano == null) null else "PROT-$ano-%"
        return jpaRepository.findMinhas(solicitanteId, estado, tipoCodigo, anoPrefixo, pageable)
            .map(SolicitacaoJpaEntity::toDomainSemEventos)
    }

    private fun carregar(id: UUID): Optional<Solicitacao> =
        jpaRepository.findById(id).map { entity -> entity.toDomain(eventosDe(id)) }

    private fun eventosDe(solicitacaoId: UUID) =
        eventoRepository.findBySolicitacaoIdOrderByCreatedAtDesc(solicitacaoId)
            .map(SolicitacaoEventoJpaEntity::toDomain)
}
